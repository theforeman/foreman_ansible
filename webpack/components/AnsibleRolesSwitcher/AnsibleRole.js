import React from 'react';
import { ListView, Icon } from 'patternfly-react';
import classNames from 'classnames';

import './styles.scss';

const clickHandler = (onClick, role) => (event) => {
  event.preventDefault();
  onClick(role);
};


const actionButton = (icon, onClick, role) => (
  <button
    href='#'
    onClick={clickHandler(onClick, role)}
    className="role-add-remove-btn"
  >
    <Icon className='fa-2x' type='fa' name={icon} />
  </button>
);

const AnsibleRole = ({ role, icon, onClick }) => (
  <ListView.Item
    id={role.id}
    className={classNames('listViewItem--listItemVariants', { 'ansible-role-disabled': role.inherited })}
    heading={role.name}
    actions={role.inherited ? '' : actionButton(icon, onClick, role)}
    stacked
  />
);

export default AnsibleRole;
