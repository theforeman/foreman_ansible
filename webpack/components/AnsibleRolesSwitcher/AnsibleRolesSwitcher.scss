@import '~@theforeman/vendor/scss/variables';

#ansibleRolesSwitcher {
  .list-view-pf {
    .list-group-item.ansible-role-disabled {
      background-color: $color-pf-black-200;
      color: $color-pf-black-500;
      border-left-color: $color-pf-black-200;
      border-right-color: $color-pf-black-200;
    }

    .list-group-item.ansible-role-movable:hover {
      cursor: pointer;
    }
  }

  .loading-state-pf {
    padding-top: 5%;
    background-color: $color-pf-white;
  }

  .role-add-remove-btn {
    background-color: initial;
    border: none;
    color: #0388ce;
  }

  .list-view-pf-main-info {
    padding-top: 10px;
    padding-bottom: 10px;
  }

  .list-view-pf-actions {
    margin-top: 0;
    margin-bottom: 0;
  }

  .list-group-item-heading {
    margin-bottom: 0;
    font-size: 12px;
    font-weight: normal;
  }
}
