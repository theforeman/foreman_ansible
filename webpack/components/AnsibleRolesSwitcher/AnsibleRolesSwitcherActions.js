import { snakeCase, camelCase } from 'lodash';
import api from 'foremanReact/API';

import {
  ANSIBLE_ROLES_REQUEST,
  ANSIBLE_ROLES_SUCCESS,
  ANSIBLE_ROLES_FAILURE,
  ANSIBLE_ROLES_ADD,
  ANSIBLE_ROLES_REMOVE,
  ANSIBLE_ROLES_ASSIGNED_PAGE_CHANGE,
} from './AnsibleRolesSwitcherConstants';

export const getAnsibleRoles = (
  url,
  initialAssignedRoles,
  inheritedRoleIds,
  resourceId,
  resourceName,
  pagination,
  search,
) => (dispatch) => {
  dispatch({ type: ANSIBLE_ROLES_REQUEST });

  const params = {
    ...propsToSnakeCase(pagination || {}),
    ...(search || {}),
    ...propsToSnakeCase({ resourceId, resourceName }),
  };

  return api.get(url, {}, params)
    .then(({ data }) => dispatch({
      type: ANSIBLE_ROLES_SUCCESS,
      payload: {
        initialAssignedRoles,
        inheritedRoleIds,
        ...propsToCamelCase(data),
      },
    }))
    .catch(error => dispatch(errorHandler(ANSIBLE_ROLES_FAILURE, error)));
};

const errorHandler = (msg, err) => {
  const error = { errorMsg: 'Failed to fetch Ansible Roles from server.', statusText: err.response.statusText };
  return ({ type: msg, payload: { error } });
};

//  stolen from katello/webpack/services/index.js and modified
const propsToSnakeCase = ob =>
  propsToCase(snakeCase, 'propsToSnakeCase only takes objects', ob);

const propsToCamelCase = ob =>
  propsToCase(camelCase, 'propsToCamelCase only takes objects', ob);

const propsToCase = (casingFn, errorMsg, ob) => {
  if (typeof (ob) !== 'object') throw Error(errorMsg);

  return Object.keys(ob).reduce((memo, key) => {
    // eslint-disable-next-line no-param-reassign
    memo[casingFn(key)] = ob[key];
    return memo;
  }, {});
};

export const addAnsibleRole = role =>
  ({ type: ANSIBLE_ROLES_ADD, payload: { role } });

export const removeAnsibleRole = role =>
  ({ type: ANSIBLE_ROLES_REMOVE, payload: { role } });

export const changeAssignedPage = pagination =>
  ({ type: ANSIBLE_ROLES_ASSIGNED_PAGE_CHANGE, payload: { pagination } });
