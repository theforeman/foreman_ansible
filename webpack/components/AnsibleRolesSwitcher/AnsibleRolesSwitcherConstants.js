export const ANSIBLE_ROLES_REQUEST = 'ANSIBLE_ROLES_REQUEST';
export const ANSIBLE_ROLES_SUCCESS = 'ANSIBLE_ROLES_SUCCESS';
export const ANSIBLE_ROLES_FAILURE = 'ANSIBLE_ROLES_FAILURE';
export const ANSIBLE_ROLES_ADD = 'ANSIBLE_ROLES_ADD';
export const ANSIBLE_ROLES_REMOVE = 'ANSIBLE_ROLES_REMOVE';
export const ANSIBLE_ROLES_ASSIGNED_PAGE_CHANGE =
  'ANSIBLE_ROLES_ASSIGNED_PAGE_CHANGE';
export const ANSIBLE_ROLES_FORM_OBJECT = 'ANSIBLE_ROLES_FORM_OBJECT';
export const ANSIBLE_VARIABLES_REQUEST = 'ANSIBLE_VARIABLES_REQUEST';
export const ANSIBLE_VARIABLES_SUCCESS = 'ANSIBLE_VARIABLES_SUCCESS';
export const ANSIBLE_VARIABLES_FAILURE = 'ANSIBLE_VARIABLES_FAILURE';
export const ANSIBLE_VARIABLES_REMOVE = 'ANSIBLE_VARIABLES_REMOVE';
