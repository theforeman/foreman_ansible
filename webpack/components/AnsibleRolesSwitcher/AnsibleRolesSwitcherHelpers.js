export const excludeAssignedRolesSearch = (assignedRoles) => {
  const searchString = assignedRoles.length === 0 ? '' : `id !^ (${assignedRoles.map(role => role.id).join(', ')})`;
  return ({ search: searchString });
};
