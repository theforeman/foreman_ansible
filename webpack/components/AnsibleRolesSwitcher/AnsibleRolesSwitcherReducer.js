import Immutable from 'seamless-immutable';

import {
  ANSIBLE_ROLES_REQUEST,
  ANSIBLE_ROLES_SUCCESS,
  ANSIBLE_ROLES_FAILURE,
  ANSIBLE_ROLES_ADD,
  ANSIBLE_ROLES_REMOVE,
  ANSIBLE_ROLES_ASSIGNED_PAGE_CHANGE,
} from './AnsibleRolesSwitcherConstants';

const ansibleRolesSuccess = (state, payload) => {
  const {
    page,
    perPage,
    subtotal,
    results,
    initialAssignedRoles,
    inheritedRoleIds,
  } = payload;

  return state.merge({
    loading: false,
    itemCount: Number(subtotal),
    pagination: { page: Number(page), perPage: Number(perPage) },
    results,
    assignedRoles: initialAssignedRoles,
    inheritedRoleIds,
  });
};

const addItem = (list, item) => ([...(list || []), item]);

const removeItem = (list, item) => list.filter(listItem => item.id !== listItem.id);

const removeItemNewState = (state, role) => ({
  assignedRoles: removeItem(state.assignedRoles, role),
  results: addItem(state.results, role),
  itemCount: state.itemCount + 1,
});

const addItemNewState = (state, role) => ({
  assignedRoles: addItem(state.assignedRoles, role),
  itemCount: state.itemCount - 1,
});

const ansibleRoleAdd = (state, payload) =>
  state.merge(addItemNewState(state, payload.role));

const ansibleRoleRemove = (state, payload) =>
  state.merge(removeItemNewState(state, payload.role));

export const initialState = Immutable({
  loading: false,
  itemCount: 0,
  pagination: {
    page: 1,
    perPage: 20,
  },
  assignedRoles: [],
  inheritedRoleIds: [],
  results: [],
  assignedPagination: {
    page: 1,
    perPage: 20,
  },
  error: { errorMsg: '', status: '', statusText: '' },
});

const ansibleRoles = (state = initialState, action) => {
  const { payload } = action;

  switch (action.type) {
    case ANSIBLE_ROLES_REQUEST:
      return state.set('loading', true);
    case ANSIBLE_ROLES_SUCCESS:
      return ansibleRolesSuccess(state, payload);
    case ANSIBLE_ROLES_FAILURE:
      return state.merge({ error: payload.error, loading: false });
    case ANSIBLE_ROLES_ADD:
      return ansibleRoleAdd(state, payload);
    case ANSIBLE_ROLES_REMOVE:
      return ansibleRoleRemove(state, payload);
    case ANSIBLE_ROLES_ASSIGNED_PAGE_CHANGE:
      return state.set('assignedPagination', payload.pagination);
    default:
      return state;
  }
};

export default ansibleRoles;
