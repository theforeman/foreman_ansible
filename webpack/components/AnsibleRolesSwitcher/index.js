import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

import AnsibleRolesSwitcher from './AnsibleRolesSwitcher';
import * as AnsibleRolesSwitcherActions from './AnsibleRolesSwitcherActions';
import AnsiblePermissionDenied from './components/AnsiblePermissionDenied';
import withProtectedView from './components/withProtectedView';
import {
  selectUnassignedRoles,
  selectAssignedRolesPage,
  selectAssignedRolesCount,
  selectResults,
  selectPaginationMemoized,
  selectItemCount,
  selectLoading,
  selectAssignedPagination,
  selectError,
} from './AnsibleRolesSwitcherSelectors';


const mapStateToProps = state => ({
  results: selectResults(state),
  pagination: selectPaginationMemoized(state),
  itemCount: selectItemCount(state),
  loading: selectLoading(state),
  error: selectError(state),
  assignedPagination: selectAssignedPagination(state),
  assignedRolesCount: selectAssignedRolesCount(state),
  assignedRoles: selectAssignedRolesPage(state),
  unassignedRoles: selectUnassignedRoles(state),
});

const mapDispatchToProps = dispatch =>
  bindActionCreators(AnsibleRolesSwitcherActions, dispatch);

export default withProtectedView(
  connect(mapStateToProps, mapDispatchToProps)(AnsibleRolesSwitcher),
  AnsiblePermissionDenied,
  props => (props.data && props.data.canView),
);
