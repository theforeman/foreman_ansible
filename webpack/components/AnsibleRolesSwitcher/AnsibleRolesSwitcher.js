import React from 'react';
import { Grid, Row, Col } from 'patternfly-react';
import { lowerCase } from 'lodash';

import AvailableRolesList from './components/AvailableRolesList';
import AssignedRolesList from './components/AssignedRolesList';
import AnsibleRolesSwitcherError from './components/AnsibleRolesSwitcherError';
import { excludeAssignedRolesSearch } from './AnsibleRolesSwitcherHelpers';

class AnsibleRolesSwitcher extends React.Component {
  componentDidMount() {
    const {
      initialAssignedRoles,
      availableRolesUrl,
      inheritedRoleIds,
      resourceId,
      resourceName,
    } = this.props.data;

    this.props.getAnsibleRoles(
      availableRolesUrl,
      initialAssignedRoles,
      inheritedRoleIds,
      resourceId,
      resourceName,
      {},
      excludeAssignedRolesSearch(initialAssignedRoles),
    );
  }

  render() {
    const {
      loading,
      pagination,
      itemCount,
      addAnsibleRole,
      removeAnsibleRole,
      getAnsibleRoles,
      changeAssignedPage,
      assignedPagination,
      assignedRolesCount,
      assignedRoles,
      unassignedRoles,
      error,
    } = this.props;

    const {
      availableRolesUrl,
      inheritedRoleIds,
      resourceId,
      resourceName,
    } = this.props.data;

    const onListingChange = paginationArgs =>
      getAnsibleRoles(
        availableRolesUrl,
        assignedRoles,
        inheritedRoleIds,
        resourceId,
        resourceName,
        paginationArgs,
        excludeAssignedRolesSearch(assignedRoles),
      );

    return (
      <Grid bsClass="container-fluid" id="ansibleRolesSwitcher">
          <Row className="row-eq-height">
            <AnsibleRolesSwitcherError error={error} />
            <Col sm={6} className="available-roles-container">
              <div className="available-roles-header">
                <h2>{__('Available Ansible Roles')}</h2>
              </div>
               <AvailableRolesList unassignedRoles={unassignedRoles}
                                   pagination={pagination}
                                   itemCount={itemCount}
                                   onListingChange={onListingChange}
                                   onAddRole={addAnsibleRole}
                                   loading={loading} />
            </Col>

            <Col sm={6} className="assigned-roles-container">
              <div className="assigned-roles-header">
                <h2>{__('Assigned Ansible Roles')}</h2>
              </div>
                 <AssignedRolesList assignedRoles={assignedRoles}
                                    pagination={assignedPagination}
                                    itemCount={assignedRolesCount}
                                    onPaginationChange={changeAssignedPage}
                                    onRemoveRole={removeAnsibleRole}
                                    resourceName={lowerCase(resourceName || '')} />
            </Col>
          </Row>
      </Grid>
    );
  }
}

export default AnsibleRolesSwitcher;
