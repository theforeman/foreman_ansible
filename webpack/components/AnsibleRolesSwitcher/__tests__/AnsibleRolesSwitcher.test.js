import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import AnsibleRolesSwitcher from '../AnsibleRolesSwitcher';

jest.mock('foremanReact/components/Pagination/PaginationWrapper');

describe('AnsibleRolesSwitcher', () => {
  const noop = () => {};

  it('should render', async () => {
    const page = shallow(<AnsibleRolesSwitcher
      loading={false}
      pagination={{ page: 1, perPage: 12 }}
      itemCount={20}
      addAnsibleRole={noop}
      removeAnsibleRole={noop}
      getAnsibleRoles={noop}
      changeAssignedPage={noop}
      assignedPagination={{ page: 1, perPage: 12 }}
      assignedRolesCount={2}
      initialAssignedRoles={[]}
      assignedRoles={[]}
      unassignedRoles={[]}
      resourceName='host'
      error={{ statusText: '', errorMsg: '' }}
    />);

    expect(toJson(page)).toMatchSnapshot();
  });
});
