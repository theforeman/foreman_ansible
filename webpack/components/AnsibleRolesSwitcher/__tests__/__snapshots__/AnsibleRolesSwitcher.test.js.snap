// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnsibleRolesSwitcher should render 1`] = `
<Grid
  bsClass="container-fluid"
  componentClass="div"
  fluid={false}
  id="ansibleRolesSwitcher"
>
  <Row
    bsClass="row"
    className="row-eq-height"
    componentClass="div"
  >
    <AnsibleRolesSwitcherError
      error={
        Object {
          "errorMsg": "",
          "statusText": "",
        }
      }
    />
    <Col
      bsClass="col"
      className="available-roles-container"
      componentClass="div"
      sm={6}
    >
      <div
        className="available-roles-header"
      >
        <h2>
          Available Ansible Roles
        </h2>
      </div>
      <AvailableRolesList
        itemCount={20}
        loading={false}
        onAddRole={[Function]}
        onListingChange={[Function]}
        pagination={
          Object {
            "page": 1,
            "perPage": 12,
          }
        }
        unassignedRoles={Array []}
      />
    </Col>
    <Col
      bsClass="col"
      className="assigned-roles-container"
      componentClass="div"
      sm={6}
    >
      <div
        className="assigned-roles-header"
      >
        <h2>
          Assigned Ansible Roles
        </h2>
      </div>
      <AssignedRolesList
        assignedRoles={Array []}
        itemCount={2}
        onPaginationChange={[Function]}
        onRemoveRole={[Function]}
        pagination={
          Object {
            "page": 1,
            "perPage": 12,
          }
        }
        resourceName=""
      />
    </Col>
  </Row>
</Grid>
`;
